File: backend/app/schemas/theme.py
"""
Theme Pydantic schemas.
"""

from datetime import datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field


class ThemeBase(BaseModel):
    name: str
    font_family: str = "Montserrat"
    font_size: int = 80
    font_weight: int = 800
    text_color: str = "#FFFFFF"
    highlight_color: str = "#FFFF00"
    outline_color: str = "#000000"
    shadow_color: str = "#000000"
    background_color: Optional[str] = None
    position_x: int = 50
    position_y: int = 70
    alignment: str = "center"
    outline_width: int = 4
    shadow_offset: int = 2
    shadow_blur: int = 0
    animation_style: str = "karaoke"
    animation_speed: float = 1.0
    words_per_line: int = 3
    max_chars_per_line: int = 30
    uppercase: bool = False


class ThemeCreate(ThemeBase):
    pass


class ThemeUpdate(BaseModel):
    name: Optional[str] = None
    font_family: Optional[str] = None
    font_size: Optional[int] = None
    text_color: Optional[str] = None
    highlight_color: Optional[str] = None
    animation_style: Optional[str] = None


class ThemeResponse(ThemeBase):
    model_config = ConfigDict(from_attributes=True)
    id: str
    is_default: bool
    is_custom: bool
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/theme.ts
export type AnimationStyle = 
  | 'none' 
  | 'karaoke' 
  | 'bounce' 
  | 'pop' 
  | 'glow' 
  | 'wave'

export type TextAlignment = 'left' | 'center' | 'right'

export type EmojiStyle = 'auto' | 'animated' | 'static' | 'none'

export interface Theme {
  id: string
  name: string
  is_default: boolean
  is_custom: boolean
  user_id: string | null
  
  // Font Settings
  font_family: string
  font_size: number
  font_weight: number
  
  // Colors
  text_color: string
  highlight_color: string
  outline_color: string
  shadow_color: string
  background_color: string | null
  
  // Position
  position_x: number // percentage
  position_y: number // percentage
  alignment: TextAlignment
  
  // Effects
  outline_width: number
  shadow_offset: number
  shadow_blur: number
  
  // Animation
  animation_style: AnimationStyle
  words_per_line: number
  max_chars_per_line: number
  
  // Emoji
  show_emojis: boolean
  emoji_style: EmojiStyle
}

export interface ThemeCreate {
  name: string
  font_family?: string
  font_size?: number
  font_weight?: number
  text_color?: string
  highlight_color?: string
  outline_color?: string
  shadow_color?: string
  background_color?: string | null
  position_x?: number
  position_y?: number
  alignment?: TextAlignment
  outline_width?: number
  shadow_offset?: number
  shadow_blur?: number
  animation_style?: AnimationStyle
  words_per_line?: number
  max_chars_per_line?: number
  show_emojis?: boolean
  emoji_style?: EmojiStyle
}

export interface ThemeUpdate extends Partial<ThemeCreate> {}

// Default themes from the architecture
export const DEFAULT_THEMES: Partial<Theme>[] = [
  {
    name: 'Hormozi',
    font_family: 'Montserrat',
    font_size: 80,
    font_weight: 800,
    text_color: '#FFFFFF',
    highlight_color: '#FFFF00',
    outline_color: '#000000',
    outline_width: 4,
    position_y: 70,
    animation_style: 'karaoke',
    words_per_line: 3,
  },
  {
    name: 'Beast',
    font_family: 'Impact',
    font_size: 90,
    font_weight: 700,
    text_color: '#FFFFFF',
    highlight_color: '#FF0000',
    outline_color: '#000000',
    outline_width: 6,
    position_y: 80,
    animation_style: 'pop',
    words_per_line: 2,
  },
  {
    name: 'Clean',
    font_family: 'Inter',
    font_size: 60,
    font_weight: 600,
    text_color: '#FFFFFF',
    highlight_color: '#00BFFF',
    outline_color: '#000000',
    outline_width: 2,
    position_y: 85,
    animation_style: 'none',
    words_per_line: 5,
  },
  {
    name: 'Neon',
    font_family: 'Poppins',
    font_size: 70,
    font_weight: 700,
    text_color: '#00FF00',
    highlight_color: '#FF00FF',
    outline_color: '#000000',
    outline_width: 3,
    shadow_blur: 10,
    position_y: 75,
    animation_style: 'glow',
    words_per_line: 3,
  },
]
